Just tested this on a local container environment with corresponding frontend PR.

* Editing chapters in the editor works! vtt-assets have been created and seem to be fine.
* When running the built-in publish-workflow from the editor, I got an error running the WOH `timelinepreviews` (in `partial-publish.yaml`)

 ```
 2025-10-31T08:28:49,129 | INFO  | (TimelinePreviewsServiceImpl:406) - Create timeline preview images file for track '145b92ee-7589-45db-a5ab-ee356cfcf747' at /data/opencast/workspace/mediapackage/5853c6e6-350b-47e5-9213-eb270a277a17/145b92ee-7589-45db-a5ab-ee356cfcf747/presentation-145b92ee-7589-45db-a5ab-ee356cfcf747_b12f247f-270a-4c56-82a1-0bdd9060572b_timelinepreviews.png
2025-10-31T08:28:49,132 | ERROR | (AbstractJobProducer$JobRunner:342) - Error handling operation 'TimelinePreview':
org.opencastproject.serviceregistry.api.ServiceRegistryException: Error handling operation 'TimelinePreview'
	at org.opencastproject.timelinepreviews.ffmpeg.TimelinePreviewsServiceImpl.process(TimelinePreviewsServiceImpl.java:355) ~[?:?]
	at org.opencastproject.job.api.AbstractJobProducer$JobRunner.call(AbstractJobProducer.java:312) ~[?:?]
	at org.opencastproject.job.api.AbstractJobProducer$JobRunner.call(AbstractJobProducer.java:271) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1575) [?:?]
Caused by: org.opencastproject.timelinepreviews.api.TimelinePreviewsException: Element is not a video track
	at org.opencastproject.timelinepreviews.ffmpeg.TimelinePreviewsServiceImpl.generatePreviewImages(TimelinePreviewsServiceImpl.java:278) ~[?:?]
	at org.opencastproject.timelinepreviews.ffmpeg.TimelinePreviewsServiceImpl.process(TimelinePreviewsServiceImpl.java:345) ~[?:?]
	... 6 more
```

* Furthermore, the WOH `analyze-tracks` seems to mistake chapters for subtitles? I'm getting `presentation_work_subtitle=true` although there are no subtitles in the mediapackage. Or is this the intended behaviour?

* The chapters-vtt are not published in engage? I can't find any vtt-files in `episode.json`. (tested with the built-in workflows/changes)

--- a/etc/workflows/partial-publish.yaml
+++ b/etc/workflows/partial-publish.yaml
@@ -137,6 +137,14 @@ operations:
       - target-flavor: captions/delivery
       - target-tags: +engage-download
 
+  # Mark chapters for publication
+  - id: tag
+    description: Mark chapters for publication
+    configurations:
+      - source-flavors: chapters/shifted
+      - target-flavor: chapters/delivery
+      - target-tags: +engage-download
+
   # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   # Encode for publication to Engage
 

--- a/etc/workflows/partial-theming.yaml
+++ b/etc/workflows/partial-theming.yaml
@@ -27,6 +27,14 @@ operations:
 
   - id: tag
     if: NOT(${theme_active})
+    description: If no shifting was necessary, mark the trimmed chapters as shifted
+    configurations:
+      - source-flavors: 'chapters/trimmed'
+      - target-flavor: 'chapters/shifted'
+      - copy: false
+
+  - id: tag
+    if: NOT(${theme_active})
     description: Tag elements as themed
     configurations:
       - source-flavors: '*/trimmed'
@@ -103,3 +111,19 @@ operations:
     configurations:
       - source-flavors: captions/trimmed
       - target-flavor: captions/shifted
+
+  - id: subtitle-timeshift
+    if: ${theme_bumper_active}
+    exception-handler-workflow: partial-error
+    description: Create new chapters with time offset
+    configurations:
+      - subtitle-source-flavor: chapters/trimmed
+      - target-flavor: chapters/shifted
+      - video-source-flavor: branding/bumper
+
+  - id: tag
+    if: NOT ${theme_bumper_active}
+    description: If no shifting was necessary, mark the trimmed chapters as shifted
+    configurations:
+      - source-flavors: chapters/trimmed
+      - target-flavor: chapters/shifted

--- a/modules/editor-service-api/src/main/java/org/opencastproject/editor/api/EditingData.java
+++ b/modules/editor-service-api/src/main/java/org/opencastproject/editor/api/EditingData.java
@@ -62,14 +62,15 @@ public class EditingData {
 
   private final List<String> waveformURIs;
   private final List<Subtitle> subtitles;
+  private final List<Subtitle> chapters;
   private final Boolean local;
 
   private final String metadataJSON;
 
   public EditingData(List<SegmentData> segments, List<TrackData> tracks, List<WorkflowData> workflows, Long duration,
           String title, String recordingStartDate, String seriesId, String seriesName, Boolean workflowActive,
-          List<String> waveformURIs, List<Subtitle> subtitles, Boolean local, Boolean lockingActive,
-          Integer lockRefresh, User user, String metadataJSON) {
+          List<String> waveformURIs, List<Subtitle> subtitles, List<Subtitle> chapters, Boolean local,
+          Boolean lockingActive, Integer lockRefresh, User user, String metadataJSON) {
     this.segments = segments;
     this.tracks = tracks;
     this.workflows = workflows;
@@ -80,6 +81,7 @@ public class EditingData {
     this.workflowActive = workflowActive;
     this.waveformURIs = waveformURIs;
     this.subtitles = subtitles;
+    this.chapters = chapters;
     this.local = local;
     this.lockingActive = lockingActive;
     this.lockRefresh = lockRefresh * 1000;
@@ -119,6 +121,10 @@ public class EditingData {
 
   public List<Subtitle> getSubtitles() {
     return subtitles;
+  }
+
+  public List<Subtitle> getChapters() {
+    return chapters;
   }
 
   public String getMetadataJSON() {

--- a/modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java
+++ b/modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java
@@ -172,6 +172,7 @@ public class EditorServiceImpl implements EditorService {
   private MediaPackageElementFlavor smilSilenceFlavor;
   private ElasticsearchIndex searchIndex;
   private MediaPackageElementFlavor captionsFlavor;
+  private MediaPackageElementFlavor chapterFlavor;
   private String thumbnailWfProperty;
   private List<MediaPackageElementFlavor> thumbnailSourcePrimary;
   private String distributionDirectory;
@@ -185,6 +186,7 @@ public class EditorServiceImpl implements EditorService {
   private static final String DEFAULT_SMIL_SILENCE_FLAVOR = "*/silence";
   private static final String DEFAULT_PREVIEW_VIDEO_SUBTYPE = "video+preview";
   private static final String DEFAULT_CAPTIONS_FLAVOR = "captions/*";
+  private static final String DEFAULT_CHAPTER_FLAVOR = "chapters/*";
   private static final String DEFAULT_THUMBNAIL_SUBTYPE = "player+preview";
   private static final String DEFAULT_THUMBNAIL_WF_PROPERTY = "thumbnail_edited";
   private static final List<MediaPackageElementFlavor> DEFAULT_THUMBNAIL_PRIORITY_FLAVOR = new ArrayList<>();
@@ -199,6 +201,7 @@ public class EditorServiceImpl implements EditorService {
   public static final String OPT_SMIL_SILENCE_FLAVOR = "smil.silence.flavor";
   public static final String OPT_PREVIEW_VIDEO_SUBTYPE = "preview.video.subtype";
   public static final String OPT_CAPTIONS_FLAVOR = "captions.flavor";
+  public static final String OPT_CHAPTER_FLAVOR = "chapter.flavor";
   public static final String OPT_THUMBNAILSUBTYPE = "thumbnail.subtype";
   public static final String OPT_THUMBNAIL_WF_PROPERTY = "thumbnail.workflow.property";
   public static final String OPT_THUMBNAIL_PRIORITY_FLAVOR = "thumbnail.priority.flavor";
@@ -339,6 +342,11 @@ public class EditorServiceImpl implements EditorService {
     captionsFlavor = MediaPackageElementFlavor.parseFlavor(
             StringUtils.defaultString((String) properties.get(OPT_CAPTIONS_FLAVOR), DEFAULT_CAPTIONS_FLAVOR));
     logger.debug("Caption flavor set to '{}'", captionsFlavor);
+
+    // Flavor for chapters
+    chapterFlavor = MediaPackageElementFlavor.parseFlavor(
+        StringUtils.defaultString((String) properties.get(OPT_CHAPTER_FLAVOR), DEFAULT_CHAPTER_FLAVOR));
+    logger.debug("Chapter flavor set to '{}'", chapterFlavor);
 
     thumbnailSubType =  Objects.toString(properties.get(OPT_THUMBNAILSUBTYPE), DEFAULT_THUMBNAIL_SUBTYPE);
     logger.debug("Thumbnail subtype set to '{}'", thumbnailSubType);
@@ -541,8 +549,12 @@ public class EditorServiceImpl implements EditorService {
    *          the subtitles to be added
    * @throws IOException
    */
-  private MediaPackage processSubtitleTrack(MediaPackage mediaPackage, List<EditingData.Subtitle> subtitles)
-          throws IOException, IllegalArgumentException {
+  private MediaPackage processSubtitleTrack(
+      MediaPackage mediaPackage,
+      List<EditingData.Subtitle> subtitles,
+      MediaPackageElementFlavor newTrackFlavor,
+      String newFileName
+  ) throws IOException, IllegalArgumentException {
     for (EditingData.Subtitle subtitle : subtitles) {
       // Generate ID for new tracks
       String subtitleId = UUID.randomUUID().toString();
@@ -551,7 +563,7 @@ public class EditorServiceImpl implements EditorService {
       // Check if subtitle already exists
       for (Track t : mediaPackage.getTracks()) {
         if (t.getIdentifier().matches(subtitle.getId())) {
-          logger.debug("Set Identifier for Subtitle-Track to: {}", t.getIdentifier());
+          logger.debug("Set Identifier for {}-Track to: {}", newTrackFlavor.getType(), t.getIdentifier());
           subtitleId = t.getIdentifier();
           trackId = t.getIdentifier();
           break;
@@ -576,14 +588,14 @@ public class EditorServiceImpl implements EditorService {
 
       // Put updated filename in working file repository and update the track.
       try (InputStream is = IOUtils.toInputStream(subtitle.getSubtitle(), "UTF-8")) {
-        URI subtitleUri = workspace.put(mediaPackage.getIdentifier().toString(), subtitleId, "subtitle.vtt", is);
+        URI subtitleUri = workspace.put(mediaPackage.getIdentifier().toString(), subtitleId, newFileName + ".vtt", is);
 
         // If not exists, create new Track
         if (track == null) {
           // TODO: Figure out which flavor new subtitles from the editor should have
           track = (Track) mediaPackage.add(subtitleUri, MediaPackageElement.Type.Track,
-              new MediaPackageElementFlavor(captionsFlavor.getType(),"source"));
-          logger.info("Creating new subtitle track {}", track.getIdentifier());
+              new MediaPackageElementFlavor(newTrackFlavor.getType(),"source"));
+          logger.info("Creating new {} track {}", newTrackFlavor.getType(), track.getIdentifier());
         }
 
         track.setURI(subtitleUri);
@@ -990,6 +1002,19 @@ public class EditorServiceImpl implements EditorService {
       }
     }
 
+    // Get chapters too
+    Track[] chapterTracks = mp.getTracks(chapterFlavor);
+    List<EditingData.Subtitle> chapters = new ArrayList<>();
+    for (Track t: chapterTracks) {
+      try {
+        File chapterFile = workspace.get(t.getURI());
+        String chapterString = FileUtils.readFileToString(chapterFile, StandardCharsets.UTF_8);
+        chapters.add(new EditingData.Subtitle(t.getIdentifier(), chapterString, t.getTags()));
+      } catch (NotFoundException | IOException e) {
+        errorExit("Could not read chapter from file", mediaPackageId, ErrorStatus.UNKNOWN);
+      }
+    }
+
     // Get tracks from the internal publication because it is a lot faster than getting them from the asset manager
     // for some reason.
     final List<TrackData> tracks = trackList.stream().map(track -> {
@@ -1046,8 +1071,8 @@ public class EditorServiceImpl implements EditorService {
     User user = securityService.getUser();
 
     return new EditingData(segments, tracks, workflows, mp.getDuration(), mp.getTitle(), event.getRecordingStartDate(),
-            event.getSeriesId(), event.getSeriesName(), workflowActive, waveformList, subtitles, localPublication,
-            lockingActive, lockRefresh, user, "");
+            event.getSeriesId(), event.getSeriesName(), workflowActive, waveformList, subtitles, chapters,
+            localPublication, lockingActive, lockRefresh, user, "");
   }
 
 
@@ -1133,7 +1158,17 @@ public class EditorServiceImpl implements EditorService {
 
     try {
       if (editingData.getSubtitles() != null) {
-        mediaPackage = processSubtitleTrack(mediaPackage, editingData.getSubtitles());
+        mediaPackage = processSubtitleTrack(mediaPackage, editingData.getSubtitles(), captionsFlavor, "subtitle");
+      }
+    } catch (IOException e) {
+      errorExit("Unable to add subtitle track to archive", mediaPackageId, ErrorStatus.UNKNOWN, e);
+    } catch (IllegalArgumentException e) {
+      errorExit("Illegal subtitle given", mediaPackageId, ErrorStatus.UNKNOWN, e);
+    }
+
+    try {
+      if (editingData.getChapters() != null) {
+        mediaPackage = processSubtitleTrack(mediaPackage, editingData.getChapters(), chapterFlavor, "chapters");
       }
     } catch (IOException e) {
       errorExit("Unable to add subtitle track to archive", mediaPackageId, ErrorStatus.UNKNOWN, e);

